danspeech/deepspeech/model.py

```diff
@@ -0,0 +1,358 @@
+        self.num_directions = 2 if bidirectional else 1
+
+    def flatten_parameters(self):
+        self.rnn.flatten_parameters()
+
+    def forward(self, x, output_lengths):
+        if self.batch_norm is not None:
+            x = self.batch_norm(x)
+        x = nn.utils.rnn.pack_padded_sequence(x, output_lengths)
+        x, h = self.rnn(x)
+        x, _ = nn.utils.rnn.pad_packed_sequence(x)
+        if self.bidirectional:
+            x = x.view(x.size(0), x.size(1), 2, -1).sum(2).view(x.size(0), x.size(1), -1)  # (TxNxH*2) -> (TxNxH) by sum
+        return x
+
+
+class Lookahead(nn.Module):
+    # Wang et al 2016 - Lookahead Convolution Layer for Unidirectional Recurrent Neural Networks
+    # input shape - sequence, batch, feature - TxNxH
+    # output shape - same as input
+    def __init__(self, n_features, context):
+        # should we handle batch_first=True?
+        super(Lookahead, self).__init__()
+        self.n_features = n_features
+        self.weight = Parameter(torch.Tensor(n_features, context + 1))
+        assert context > 0
+        self.context = context
+        self.register_parameter('bias', None)
+        self.init_parameters()
+
+    def init_parameters(self):  # what's a better way initialiase this layer?
+        stdv = 1. / math.sqrt(self.weight.size(1))
+        self.weight.uniform_(-stdv, stdv)
+
+    def forward(self, input):
+        seq_len = input.size(0)
+        # pad the 0th dimension (T/sequence) with zeroes whose number = context
+        # Once pytorch's padding functions have settled, should move to those.
+        padding = torch.zeros(self.context, *(input.size()[1:])).type_as(input)
+        x = torch.cat((input, padding), 0)
+
+        # add lookahead windows (with context+1 width) as a fourth dimension
+        # for each seq-batch-feature combination
+        x = [x[i:i + self.context + 1] for i in range(seq_len)]  # TxLxNxH - sequence, context, batch, feature
+        x = torch.stack(x)
+        x = x.permute(0, 2, 3, 1)  # TxNxHxL - sequence, batch, feature, context
+
+        x = torch.mul(x, self.weight).sum(dim=3)
+        return x
+
+    def __repr__(self):
+        return self.__class__.__name__ + '(' \
+               + 'n_features=' + str(self.n_features) \
+               + ', context=' + str(self.context) + ')'
+
+
+class DeepSpeech(nn.Module):
+
+    # ToDO: Rename nb_layers to nb_rnn_layers or something
+    def __init__(self, rnn_type=nn.LSTM, labels=None, rnn_hidden_size=768, nb_layers=5, audio_conf=None,
+                 bidirectional=True, context=20, conv_layers=2):
+        super(DeepSpeech, self).__init__()
+
+        if not labels:
+            with open("labels.json", "r", encoding="utf-8") as label_file:
+                labels = str(''.join(json.load(label_file)))
+
+        if audio_conf is None:
+            audio_conf = {}
+
+        # model metadata needed for serialization/deserialization
+        self.version = '0.0.1'
```

Comment:
Should we consider adding model version control in later iterations?

Reply:
I think we should ignore model version control in the pytorch model (i.e. delete it at some point) and solely rely on model version control through hashing and releases instead. :)

Reply:
Seems reasonable.